storage/innobase/trx/trx0purge.cc

```diff
@@ -554,16 +579,62 @@ dberr_t purge_sys_t::iterator::free_history() const
       ut_ad(rseg.is_persistent());
       log_free_check();
       rseg.latch.wr_lock(SRW_LOCK_CALL);
-      dberr_t err=
-        trx_purge_truncate_rseg_history(rseg, *this, !rseg.is_referenced() &&
-                                        purge_sys.sees(rseg.needs_purge));
+      dberr_t err= free_history_rseg(rseg);
       rseg.latch.wr_unlock();
       if (err)
         return err;
     }
   return DB_SUCCESS;
 }
 
+/** Start truncating an undo tablespace.
+@return undo tablespace that is chosen for truncation
+@retval nullptr  if undo tablespace truncation is not enabled or possible */
+fil_space_t *purge_sys_t::truncating_tablespace()
```

Comment:
There are two callers of this function, and both of them need to check both conditions (whether truncation is already in progress, or it can be started) as well as try to start undo tablespace truncation.

With your suggestion, multiple calls or inline code duplication would be needed at both call sites.

Reply:
I will split some of the code to two new `inline` functions `purge_sys_t::undo_truncate_try()` and `trx_sys_t::undo_truncate_start()`.